cpp/src/io/orc/writer_impl.hpp

```diff
@@ -284,24 +284,61 @@ class writer::impl {
     hostdevice_2dvector<gpu::encoder_chunk_streams>* enc_streams,
     hostdevice_2dvector<gpu::StripeStream>* strm_desc);
 
+  struct intermediate_statistics {
+    intermediate_statistics(rmm::cuda_stream_view stream) : stripe_stat_chunks(0, stream){};
+    intermediate_statistics(std::vector<ColStatsBlob> rb,
+                            rmm::device_uvector<statistics_chunk> sc,
+                            hostdevice_vector<statistics_merge_group> smg,
+                            hostdevice_vector<stats_column_desc> scd)
+      : rowgroup_blobs(std::move(rb)),
+        stripe_stat_chunks(std::move(sc)),
+        stripe_stat_merge(std::move(smg)),
+        stats_desc(std::move(scd)){};
+
+    // blobs for the rowgroup. Not persisted
+    std::vector<ColStatsBlob> rowgroup_blobs;
+
+    rmm::device_uvector<statistics_chunk> stripe_stat_chunks;
+    hostdevice_vector<statistics_merge_group> stripe_stat_merge;
+    hostdevice_vector<stats_column_desc> stats_desc;
+  };
+
+  // used for chunked writes to persist data between calls to write.
+  struct persisted_statistics {
+    std::vector<rmm::device_uvector<statistics_chunk>> stripe_stat_chunks;
+    std::vector<hostdevice_vector<statistics_merge_group>> stripe_stat_merge;
+    hostdevice_vector<stats_column_desc> stats_desc;
+  };
+
   struct encoded_statistics {
-    std::vector<ColStatsBlob> rowgroup_level;
     std::vector<ColStatsBlob> stripe_level;
     std::vector<ColStatsBlob> file_level;
   };
 
   /**
-   * @brief Returns column statistics encoded in ORC protobuf format.
+   * @brief Returns column statistics in an intermediate format.
    *
    * @param statistics_freq Frequency of statistics to be included in the output file
    * @param orc_table Table information to be written
-   * @param columns List of columns
    * @param segmentation stripe and rowgroup ranges
-   * @return The statistic blobs
+   * @return The statistic information
    */
-  encoded_statistics gather_statistic_blobs(statistics_freq statistics_freq,
-                                            orc_table_view const& orc_table,
-                                            file_segmentation const& segmentation);
+  intermediate_statistics gather_statistic_blobs(statistics_freq const statistics_freq,
+                                                 orc_table_view const& orc_table,
+                                                 file_segmentation const& segmentation);
+
+  /**
+   * @brief Returns column statistics encoded in ORC protobuf format.
+   *
+   * @param stats_freq Frequence of statistics to be included in the output file
+   * @param orc_table Table information to be written
+   * @param segmentation stripe and rowgroup ranges
+   * @param incoming_stats intermediate statistics returned from `gather_statistic_blobs`
+   * @return The encoded statistic blobs
+   */
+  encoded_statistics finish_statistic_blobs(int const num_columns,
+                                            int const num_stripes,
```

Comment:
We can. This was created before I conceded storing off the stats_desc inside the intermediate stats. I'll update that.

Reply:
Actually we can't remove `num_stripes` because we do have the number of stripe blobs, we don't know how many stripes unless there is some math there to compute it. I've removed `num_columns` and left `num_stripes` for now.